drivers/ethernet/eth_stellaris.c

```diff
@@ -100,37 +99,76 @@ static int eth_stellaris_send(struct net_if *iface, struct net_pkt *pkt)
 	/* Send the payload, if partial words are present,
 	 * send them separately.
 	 */
+	partial_data = false;
+	partial_data_len = 0;
+
 	for (frag = pkt->frags; frag; frag = frag->frags) {
 		data_ptr = frag->data; bytes_left = frag->len;
-		/* Check if partial words present */
-		if ((bytes_left & 3) == 0) {
-			/* No partial words present */
-			for (; bytes_left > 0; bytes_left -= 4, data_ptr += 4) {
-				sys_write32(*(u32_t *)data_ptr, REG_MACDATA);
-			}
-		} else {
-			/* partial words present */
-			for (; bytes_left > 3; bytes_left -= 4, data_ptr += 4) {
-				sys_write32(*(u32_t *)data_ptr, REG_MACDATA);
+		if (partial_data) {
+			partial_data = false;
+			switch (partial_data_len) {
+			case 3:
+				reg_val |= data_ptr[0] << 8;
+				reg_val |= data_ptr[1] << 16;
+				reg_val |= data_ptr[2] << 24;
+				data_ptr += 3; bytes_left += 3;
+				sys_write32(reg_val, REG_MACDATA);
+				break;
+			case 2:
+				reg_val |= data_ptr[0] << 16;
+				reg_val |= data_ptr[1] << 24;
+				data_ptr += 2; bytes_left -= 2;
+				sys_write32(reg_val, REG_MACDATA);
+				break;
+			case 1:
+				reg_val |= data_ptr[0] << 24;
+				++data_ptr; --bytes_left;
+				sys_write32(reg_val, REG_MACDATA);
+				break;
 			}
 		}
-	}
 
-	/* Write the last, partial word in net_buf */
-	if (bytes_left > 0) {
-		reg_val = 0;
-		switch (bytes_left) {
-		case 3:
-			reg_val |= ((u32_t)data_ptr[2] << 16);
-		case 2:
-			reg_val |= ((u32_t)data_ptr[1] << 8);
-		case 1:
-			reg_val |= (u32_t)data_ptr[0];
-			break;
-		default:
-			break;
+		while (bytes_left > 3) {
+			sys_write32(*(u32_t *)data_ptr, REG_MACDATA);
+			data_ptr += 4;
+			bytes_left -= 4;
+		}
+		/* Have read a full fragment, hence move to next */
+		if (bytes_left == 0) {
+			continue;
+		} else if (frag->frags) { /* Next frag exists */
+			reg_val = 0;
+			partial_data = true;
+			partial_data_len = 4 - bytes_left;
+			switch (bytes_left) {
+			case 3:
+				reg_val |= data_ptr[0];
+				reg_val |= data_ptr[1] << 8;
+				reg_val |= data_ptr[2] << 16;
+				break;
+			case 2:
+				reg_val |= data_ptr[0];
+				reg_val |= data_ptr[1] << 8;
+				break;
+			case 1:
+				reg_val |= data_ptr[0];
+				break;
+			}
+		} else { /* Next frag doesn't exist. This is the last frag */
+			reg_val = 0;
+			switch (bytes_left) {
+			case 3:
+				reg_val |= (data_ptr[2] << 16);
+			case 2:
+				reg_val |= data_ptr[1] << 8;
+			case 1:
+				reg_val |= data_ptr[0];
+				break;
+			default:
+				break;
+			}
+			sys_write32(reg_val, REG_MACDATA);
```

Comment:
This logic can be simplified as shown below:

``````
void eth_stellaris_send_byte(dev, byte)
{
	dev->tx_word |= byte << dev->tx_pos;
	dev->tx_pos++;

	if (dev->tx_pos == 4) {
		sys_write32(...);
		dev->tx_pos = 0;
		dev->tx_word = 0;
	}
}

void eth_stellaris_flush(dev)
{
	sys_write32(...);
}


for (frag = pkt->frags; frag; frag = frag->frags) {
	for (i = 0; i < frag->len; i++) {
		eth_stellaris_send_byte(dev_data, frag->data[i]);
	}
}
eth_stellaris_flush(dev_data);
``````